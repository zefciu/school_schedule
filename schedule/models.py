from django.db import models
from django.db.models import ForeignKey

from schedule.const import DayOfWeek


class Named(models.Model):
    name = models.TextField(unique=True)

    def __str__(self):
        return self.name

    class Meta:
        abstract=True



class SchoolClass(Named):
    pass


class Teacher(Named):
    pass


class Student(Named):
    school_class = models.ForeignKey(
        SchoolClass,
        related_name="students",
        on_delete=models.CASCADE,
    )


class Subject(Named):
    teacher = ForeignKey(
        Teacher,
        related_name="subjects",
        on_delete=models.CASCADE,
    )




class Schedule(models.Model):
    school_class = models.ForeignKey(SchoolClass, on_delete=models.CASCADE)
    subject = models.ForeignKey(Subject, on_delete=models.CASCADE)
    dow = models.IntegerField(choices=DayOfWeek)
    hour = models.TimeField()

    class Meta:
        constraints = [
            models.UniqueConstraint(name="class_slot", fields=["school_class", "dow", "hour"]),
            models.UniqueConstraint(name="subject_slot", fields=["subject", "dow", "hour"]),
        ]
