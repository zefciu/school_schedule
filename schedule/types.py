from typing import Protocol, Any

from django.db.models import QuerySet

from schedule.models import Schedule


class ScheduleQueryParams(Protocol):
    class_name: str
    for_today: bool


class ScheduleGetter(Protocol):
    async def __call__(self, query_params: ScheduleQueryParams) -> QuerySet[Schedule]:
        ...


class SchedulesSerializingService(Protocol):
    async def __call__(self, services: QuerySet[Schedule]) -> Any:
        ...


class ScheduleQueryParamsGetter(Protocol):
    def __call__(self, args: Any) -> ScheduleQueryParams:
        ...

class ScheduleCacheKeyGenerator(Protocol):
    async def __call__(self, params: ScheduleQueryParams) -> str:
        ...

