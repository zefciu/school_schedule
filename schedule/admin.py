from django.contrib import admin

from schedule.models import SchoolClass, Subject, Teacher, Student, Schedule

admin.site.register(SchoolClass)
admin.site.register(Subject)
admin.site.register(Teacher)
admin.site.register(Student)
admin.site.register(Schedule)

# Register your models here.
