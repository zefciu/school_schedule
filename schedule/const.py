from django.db import models


class DayOfWeek(models.IntegerChoices):
    MONDAY = 1
    TUESDAY = 2
    WEDNESDAY = 3
    THURSDAY = 4
    FRIDAY = 5
    SATURDAY = 6
    SUNDAY = 7
