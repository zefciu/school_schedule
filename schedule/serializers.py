from adrf.serializers import ModelSerializer
from rest_framework import serializers
from rest_framework.fields import IntegerField

from schedule.models import Schedule, SchoolClass, Teacher, Subject


class TeacherSerializer(ModelSerializer):
    class Meta:
        model = Teacher
        fields = ["name"]


class SubjectSerializer(ModelSerializer):
    class Meta:
        model = Subject
        fields = ["name"]


class ClassSerializer(ModelSerializer):
    class Meta:
        model = SchoolClass
        fields = ["name"]


class ScheduleSerializer(ModelSerializer):

    class Meta:
        model = Schedule
        fields = ["class", "day_of_week", "hour", "subject", "teacher"]

    day_of_week = IntegerField(source="dow")
    vars()["class"] = ClassSerializer(source="school_class")
    teacher = TeacherSerializer(source="subject.teacher")
    subject = SubjectSerializer()

    def to_representation(self, instance, *args, **kwargs):
        data = super().to_representation(instance, *args, **kwargs)
        data["class"]["student_count"] = instance.class_student_count
        return data


class QueryParamsSerializer(serializers.Serializer):
    class_name = serializers.CharField(required=True)
    for_today = serializers.BooleanField(required=False, default=False)
