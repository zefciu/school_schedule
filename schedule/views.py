# Create your views here.
from adrf.views import APIView
from django.core.cache import cache
from rest_framework.response import Response

from schedule.types import (
    ScheduleGetter,
    SchedulesSerializingService,
    ScheduleQueryParamsGetter,
    ScheduleCacheKeyGenerator,
)


class SchedulesView(APIView):

    get_schedule: ScheduleGetter | None = None
    serialize_schedules: SchedulesSerializingService | None = None
    get_query_params: ScheduleQueryParamsGetter | None = None
    get_cache_key: ScheduleCacheKeyGenerator | None = None

    async def get(
            self,
            request,

    ):
        query_params = self.get_query_params(request.query_params)
        cache_key = self.get_cache_key(query_params)
        if (data := (await cache.aget(cache_key))) is None:
            schedules = await self.get_schedule(
                query_params
            )
            data = await self.serialize_schedules(schedules)
            await cache.aset(cache_key, data)
        return Response(data)
