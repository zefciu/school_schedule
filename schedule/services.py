import datetime
from dataclasses import dataclass
from typing import Any

from asgiref.sync import sync_to_async
from django.db.models import QuerySet, Count
from django.http import QueryDict

from schedule.models import Schedule
from schedule.serializers import ScheduleSerializer, QueryParamsSerializer
from schedule.types import ScheduleQueryParams as ScheduleQueryParamsProtocol


async def get_schedule(params: ScheduleQueryParamsProtocol) -> "QuerySet[Schedule]":
    kwargs: dict[str, Any] = {"school_class__name": params.class_name}
    if params.dow:
        kwargs["dow"] = params.dow
    return Schedule.objects.filter(**kwargs)\
        .annotate(class_student_count=Count("school_class__students"))\
        .all()


async def serialize_schedules(schedules: QuerySet[Schedule]):
    return await sync_to_async(lambda: ScheduleSerializer(schedules, many=True).data)()


@dataclass
class ScheduleQueryParams(ScheduleQueryParamsProtocol):
    class_name: str
    dow: int | None


def get_params(params: QueryDict) -> ScheduleQueryParams:
    serializer = QueryParamsSerializer(data=params)
    serializer.is_valid(raise_exception=True)
    if serializer.validated_data["for_today"]:
        dow = datetime.date.today().weekday() + 1    # ISO convention
    else:
        dow = None
    return ScheduleQueryParams(
        class_name=serializer.validated_data["class_name"],
        dow=dow,
    )


def get_schedules_cache_key(params: ScheduleQueryParams) -> str:
    return f"schedules-{params.class_name}-{params.dow}"
