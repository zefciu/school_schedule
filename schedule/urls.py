from django.urls import path

from schedule.services import get_schedule, get_params, serialize_schedules, get_schedules_cache_key
from schedule.views import SchedulesView

urlpatterns = [
    path("", SchedulesView.as_view(
        get_schedule=get_schedule,
        get_query_params=get_params,
        get_cache_key=get_schedules_cache_key,
        serialize_schedules=serialize_schedules,
    ), name="schedules")
]