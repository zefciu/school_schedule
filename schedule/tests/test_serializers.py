import datetime
import pytest

from schedule.models import Schedule, SchoolClass, Teacher, Subject
from schedule.serializers import ScheduleSerializer


@pytest.fixture
def schedule() -> Schedule:
    schedule = Schedule(
        school_class=SchoolClass(
            name="5A",
        ),
        subject=Subject(
            name="Math",
            teacher=Teacher(
                name="Alex",
            ),
        ),
        dow=7,
        hour=datetime.time(hour=8)
    )
    # The next line emulates the .annotate, which works poorly with mypy
    schedule.class_student_count = 23   # type: ignore
    return schedule


def test_serialize_example(snapshot, schedule: Schedule):
    serializer = ScheduleSerializer(instance=schedule)
    snapshot.assert_match(serializer.data)
