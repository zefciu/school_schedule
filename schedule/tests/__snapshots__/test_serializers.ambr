# serializer version: 1
# name: test_serialize_example
  ReturnDict({
    'class': dict({
      'name': '5A',
      'student_count': 23,
    }),
    'day_of_week': 7,
    'hour': '08:00:00',
    'subject': dict({
      'name': 'Math',
    }),
    'teacher': dict({
      'name': 'Alex',
    }),
  })
# ---
